In [121]:
import pandas as pd
melb_data = pd.read_csv('data/melb_data.csv', sep=',')

Данные представляют собой таблицу, в которой содержится 23 столбца:

index — номер строки
Suburb — наименование пригорода
Address — адрес
Rooms — количество комнат в помещении
Type — тип здания (h — дом, коттедж, вилла, терраса; u — блочный, дуплексный дом; t — таунхаус)
Price — цена помещения
Method — метод продажи 
SellerG — риэлторская компания
Date — дата продажи (в формате день/месяц/год)
Distance — расстояния до объекта от центра Мельбурна 
Postcode — почтовый индекс
Bedroom — количество спален
Bathroom — количество ванных комнат
Car — количество парковочных мест
Landsize — площадь прилегающей территории
BuildingArea — площадь здания
YearBuilt — год постройки
CouncilArea — региональное управление
Lattitude — географическая широта
Longitude — географическая долгота
Regionname — наименование района Мельбурна
Propertycount — количество объектов недвижимости в районе, выставленных на продажу
Coordinates — широта и долгота, объединённые в кортеж

ВЫВОД ПЕРВЫХ И ПОСЛЕДНИХ СТРОК

Для этого у DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n = 5).

In [122]:
display(melb_data.head())

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079    144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093    144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969    144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072    144.9941  Northern Metropolitan         4019.0   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

In [123]:
#Следующий код выведет семь последних строк нашей таблицы:

melb_data.tail(7)

index         Suburb        Address  Rooms Type      Price Method  \
13573  13573       Werribee    5 Nuragi Ct      4    h   635000.0      S   
13574  13574    Westmeadows     9 Black St      3    h   582000.0      S   
13575  13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown    83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   

             SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
13573  hockingstuart  26/08/2017      14.7  ...  1.0     662.0         172.0   
13574            Red  26/08/2017      16.5  ...  2.0     256.0         126.0   
13575          Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576       Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577          Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578        Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579        Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
13573     1980.0          NaN  -37.89327   144.64789   
13574     1970.0          NaN  -37.67917   144.89390   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13573        Western Metropolitan        16166.0  -37.89327, 144.64789  
13574       Northern Metropolitan         2474.0   -37.67917, 144.8939  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[7 rows x 23 columns]

РАЗМЕРНОСТЬ ТАБЛИЦЫ

In [124]:
#Далее хотелось бы узнать размер таблицы — количество строк и количество столбцов. Это можно сделать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов:

melb_data.shape

(13580, 23)

ПОЛУЧЕНИЕ ИНФОРМАЦИИ О СТОЛБЦАХ

In [125]:
#Для того чтобы получить более детальную информацию о столбцах таблицы, можно использовать метод DataFrame info():

melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

Данный метод выводит:

информацию об индексах;
информацию об общем количестве столбцов;
таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
количество оперативной памяти в мегабайтах, которое тратится на хранение данных.
Из вывода метода info() становится понятно, что в нашей таблице есть столбец CouncilArea с пропущенными значениями — количество непустых значений в столбце меньше, чем количество строк в таблице (12211 < 13580).

Пустыми, или пропущенными, значениями называются значения в ячейках таблицы, которые не заполнены по какой-либо причине, то есть на их месте стоит пустое место. В Pandas такие значения обозначаются символом NaN (Not-a-Number).

→ Мы будем говорить о работе с пропусками в отдельном модуле, посвящённом очистке данных, а пока что не будем обращать на них внимания.

Если вам всё же любопытно, как можно работать с такими значениями в Pandas, то вы можете посмотреть информацию здесь. https://pythonru.com/biblioteki/not-a-number-vse-o-nan-pd-5

ПОЛУЧЕНИЕ ИНФОРМАЦИИ О СТОЛБЦАХ

Для того чтобы получить более детальную информацию о столбцах таблицы, можно использовать метод DataFrame info():
Данный метод выводит:
- информацию об индексах;
- информацию об общем количестве столбцов;
- таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце   
    и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
- количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

In [126]:


melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

Из вывода метода info() становится понятно, что в нашей таблице есть столбец CouncilArea с пропущенными значениями — количество непустых значений в столбце меньше, чем количество строк в таблице (12211 < 13580).
Пустыми, или пропущенными, значениями называются значения в ячейках таблицы, которые не заполнены по какой-либо причине, то есть на их месте стоит пустое место. В Pandas такие значения обозначаются символом NaN (Not-a-Number).

ИЗМЕНЕНИЕ ТИПА ДАННЫХ В СТОЛБЦЕ

метод astype(), который позволяет преобразовать тип данных столбца

Если присмотреться внимательнее к выводу метода info(), а конкретнее — к типам данных столбцов, становится понятно, что некоторые признаки кодируются не совсем корректными типами данных. 
Например, данные в столбцах, которые отражают количество, должны, по идее, выражаться целым числом (Car, Bedroom, Bathroom и Propertyсount), однако кодируются float64 — числом с плавающей запятой размером 64 бита.
Наконец, данные в столбце с годом постройки (YearBuilt) также представлены в формате чисел с плавающей точкой.
Чтобы исправить это, можно воспользоваться методом astype(), который позволяет преобразовать тип данных столбца:

В данном коде мы при помощи метода astype() последовательно переопределяем столбцы на них же самих, только с изменённым типом данных: int64 — целочисленное число размером 64 бита.

In [127]:
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

ДОПОЛНИТЕЛЬНО Подробнее о типах данных в Pandas можно прочитать здесь. https://dfedorov.spb.ru/pandas/Обзор%20типов%20данных%20pandas.html

ПОЛУЧЕНИЕ ОПИСАТЕЛЬНОЙ СТАТИСТИКИ

→ Часто при работе с таблицей нужно быстро посмотреть на основные статистические свойства её столбцов. Для этого можно воспользоваться методом DataFrame describe().
По умолчанию метод работает с числовыми (int64 и float64) столбцами и показывает 
число непустых значений (count), 
среднее (mean), 
стандартное отклонение (std), 
минимальное значение (min),  
квантили уровней 0.25, 0.5 (медиана) и 0.75 (25%, 50%, 75%) и 
максимальное значение (max) для каждого столбца исходной таблицы.

In [128]:
melb_data.describe().loc[:, ['Distance', 'BuildingArea' , 'Price']]

Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

Какие интересные выводы мы можем сделать, исходя из построенной таблицы?
Столбец Distance (расстояние до центра города)
В нашем наборе данных представлены объекты, удалённые от центра города на расстояние до 48 километров.

Столбец BuildingArea (площадь здания)
В нашей таблице представлены участки, на которых отсутствует само здание (его площадь равна 0), то есть владельцы продавали участок земли без строений на нём.
Большая часть объектов (75%) имеет площадь здания до 130 квадратных метров, однако в нашей таблице есть настоящий дворец (судя по его размерам) с площадью в 44 515 квадратных метров.


Столбец Price (цена)
Диапазон цен на недвижимость в Мельбурне варьируется от 85 тысяч до 9 миллионов австралийских долларов.

На самом деле метод describe() можно применять не только к числовым признакам. С помощью параметра include можно указать тип данных, для которого нужно вывести описательную информацию.

Например, для типа данных object метод describe() возвращает DataFrame, в котором указаны:

количество непустых строк (count);
количество уникальных значений (unique);
самое частое значение — мода —  (top);
частота — объём использования — этого значения (freq) для каждого столбца типа object исходной таблицы

In [129]:
melb_data.describe(include=['object'])

Suburb           Address   Type Method SellerG        Date  \
count       13580             13580  13580  13580   13580       13580   
unique        314             13378      3      5     268          58   
top     Reservoir  36 Aberfeldie St      h      S  Nelson  27/05/2017   
freq          359                 3   9449   9022    1565         473   

       CouncilArea             Regionname         Coordinates  
count        12211                  13580               13580  
unique          33                      8               13097  
top       Moreland  Southern Metropolitan  -37.8361, 144.9966  
freq          1163                   4695                  12

Также приведём несколько выводов, которые можно сделать из полученной таблицы:

Столбец Suburb (пригород)
Наибольшее количество проданных объектов (359) находилось в пригороде Reservoir.


Столбец Type (тип здания)
Самый популярный тип дома — h (дом, вилла, коттедж).

Столбец SellerG (риелтор)
В наших данных нам известно о 268 различных риэлторских компаниях, однако самой главной «акулой» в этом бизнесе является компания Nelson — они продали 1 565 различных домов.

Столбец Date (дата продажи)
В нашей таблице содержится информация за 58 дней, при этом наибольшее число продаж (473) пришлось на 27 мая 2017 года.

ПОЛУЧЕНИЕ ЧАСТОТЫ УНИКАЛЬНЫХ ЗНАЧЕНИЙ В СТОЛБЦЕ

Для того чтобы определить, сколько раз в столбце повторяется каждый из вариантов значений (т.е. найти частоту для каждого уникального значения), используется метод value_counts().
Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями — соответствующая им частота.

In [130]:
#Рассмотрим работу value_counts() на примере столбца с названиями районов:

melb_data['Regionname'].value_counts()

Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: Regionname, dtype: int64

In [131]:
#Чтобы сделать вывод более интерпретируемым и понятным, можно воспользоваться параметром normalize. 
# При установке значения этого параметра на True результат будет представляться в виде доли (относительной частоты):
melb_data['Regionname'].value_counts(normalize=True)

#Из результатов мы можем сделать вывод, 
# что наименьшее количество объектов (менее 1%) было продано в районах Victoria, а наибольшее — в районах Metropolitan.

Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: Regionname, dtype: float64

In [132]:
#Задание 6.5 Сколько пропущенных значений в столбце CouncilArea?

13580-melb_data['CouncilArea'].count()


1369

In [133]:
#Задание 6.6 Сколько столбцов после преобразования типов имеет целочисленный тип данных?
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

In [134]:
melb_data['Regionname'].describe()

count                     13580
unique                        8
top       Southern Metropolitan
freq                       4695
Name: Regionname, dtype: object

In [135]:
melb_data['Coordinates'].describe()

count                  13580
unique                 13097
top       -37.8361, 144.9966
freq                      12
Name: Coordinates, dtype: object

In [136]:
melb_data['Coordinates'].count() - melb_data['Coordinates'].nunique()

483

АГРЕГИРУЮЩИЕ МЕТОДЫ

Агрегирующим в Pandas называется метод, который для каждого столбца возвращает только одно значение — показатель (например, вычисление медианы, максимума, среднего и так далее).

МЕТОД	СТАТИСТИЧЕСКИЙ ПАРАМЕТР
.count()	Количество непустых значений
.mean()	Среднее значение
.min()	Минимальное значение
.max()	Максимальное значение
.var()	Дисперсия
.std()	Стандартное отклонение
.sum()	Сумма
.quantile(x)	Квантиль уровня x
.nunique()	Число уникальных значений
.median()  медиана

Если один из этих методов применить ко всему DataFrame, то в результате его работы будет получен объект типа Series, в котором в качестве индексов будут выступать наименования столбцов, а в качестве значений — статистический показатель. В случае применения метода к отдельному столбцу результатом вычислений станет число.


В каждый метод можно передать некоторые параметры, среди которых:

axis  — определяет, подсчитывать параметр по строкам или по столбцам;
numeric_only — определяет, вычислять параметры только по числовым столбцам/строкам или нет (True/False).

In [137]:
#Вычислим среднюю цену на объекты недвижимости:

print(melb_data['Price'].mean())

1075684.079455081


In [138]:
#Найдём максимальное количество парковочных мест:

print(melb_data['Car'].max())
# 10

10


In [139]:
#А теперь представим, что риэлторская ставка для всех компаний за продажу недвижимости составляет 12%. 
# Найдём общую прибыльность риэлторского бизнеса в Мельбурне. Результат округлим до сотых:

rate = 0.12
income = melb_data['Price'].sum() * rate
print('Total income of real estate agencies:', round(income, 2))
# Total income of real estate agencies: 1752934775.88

Total income of real estate agencies: 1752934775.88


In [140]:
#Найдём, насколько медианная площадь территории отличается от её среднего значения. 
# Вычислим модуль разницы между медианой и средним и разделим результат на среднее, чтобы получить отклонение в долях:

landsize_median = melb_data['Landsize'].median() 
landsize_mean =  melb_data['Landsize'].mean()
print(abs(landsize_median - landsize_mean)/landsize_mean*100)

#В результате получаем долю отклонения медианы от среднего значения. 
# Умножив результат на 100, получим его в процентах. Отклонение медианы от среднего значения на 21% является довольно большим, 
# и это повод задуматься над тем, чтобы исследовать признак на наличие аномалий. 

21.205713983546193


МОДА

→ Отдельный интерес представляет статический показатель моды — самого распространённого значения в столбце. Он вычисляется с помощью метода mode().

Модальных значений может быть несколько, то есть несколько значений могут встречаться одинаковое количество раз. Поэтому метод mode(), в отличие от агрегирующих методов, возвращает не одно число, а серию.

In [141]:
#Вычислим, какое число комнат чаще всего представлено на рынке недвижимости:

print(melb_data['Rooms'].mode())
# 0    3
# dtype: int64

0    3
Name: Rooms, dtype: int64


In [142]:
#Примечание. Метод mode() может быть использован не только с числовыми столбцами, но и со столбцами типа object. 
# Так, например, с помощью следующего кода можно найти наиболее распространённое название района:

melb_data['Regionname'].mode()

0    Southern Metropolitan
Name: Regionname, dtype: object

In [143]:
#Задание 7.2 Чему равно максимальное количество домов на продажу в районе (Propertycount)
melb_data['Propertycount'].max()

21650

In [144]:
#Задание 7.3 Чему равно стандартное отклонение (разброс) расстояния от центра города до объекта недвижимости?
melb_data['Distance'].std()

5.868724943071711

In [145]:
#Задание 7.4 Чему равно отклонение (в процентах) медианного значения площади здания от его среднего значения?

(melb_data['BuildingArea'].mean()- melb_data['BuildingArea'].median())/melb_data['BuildingArea'].mean()*100

9.764079662364534

In [146]:
#Задание 7.6  Сколько спален чаще всего встречается в домах в Мельбурне?

melb_data['Bedroom'].mode()

0    3
Name: Bedroom, dtype: int64

Фильтрация данных в DataFrame

Под фильтрацией в DataFrame подразумевается получение новой таблицы путём вырезания строк, не удовлетворяющих поставленному условию. 

Разберём классический способ фильтрации в DataFrame — фильтрацию с помощью масок.

Маской называется Series, которая состоит из булевых значений, при этом значения True соответствуют тем индексам, для которых заданное условие выполняется, в противном случае ставится значение False (например, цена > 2 млн).

Условия можно комбинировать, используя операторы & (логическое И) и | (логическое ИЛИ). Условия при этом заключаются в скобки.
Примечание. Обратите внимание, что использование привычных операторов and и or будет неверным и приведёт к ошибке, так как они выполняют логические операции между двумя булевыми числами. В нашем случае слева и справа от оператора стоят маски (объекты Series), для которых логическую операцию надо совершить поэлементно, а операторы and и or для такого не предназначены.

In [147]:
#Создадим маску и положим её в переменную с именем mask. Синтаксис очень прост:

mask = melb_data['Price'] > 2000000
display(mask)

0        False
1        False
2        False
3        False
4        False
         ...  
13575    False
13576    False
13577    False
13578     True
13579    False
Name: Price, Length: 13580, dtype: bool

In [148]:
#Для фильтрации нужно просто подставить переменную mask в индексацию DataFrame. 
# Маска показывает, какие строки нужно оставлять в результирующем наборе, а какие — убирать 
# (выведем первые пять строк отфильтрованной таблицы):

display(melb_data[mask].head())

index       Suburb               Address  Rooms Type      Price Method  \
80     80  Albert Park  112 Beaconsfield Pde      3    h  2850000.0     PI   
85     85  Albert Park     104 Richardson St      4    h  2300000.0      S   
88     88  Albert Park        29 Faussett St      2    h  2120000.0      S   
92     92  Albert Park           2 Dundas Pl      3    h  2615000.0      S   
93     93  Albert Park          23 Finlay St      5    h  2100000.0      S   

     SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
80    Buxton   4/03/2017       3.3  ...    0     211.0         198.0   
85  Marshall   7/05/2016       3.3  ...    1     153.0         180.0   
88      Greg  10/09/2016       3.3  ...    1     199.0         107.0   
92    Cayzer  10/12/2016       3.3  ...    1     177.0         181.0   
93      Greg  10/12/2016       3.3  ...    1     237.0         126.0   

    YearBuilt   CouncilArea  Lattitude  Longtitude             Regionname  \
80       1890  Port Phillip   -37.8481    144.9499  Southern Metropolitan   
85       1880  Port Phillip   -37.8447    144.9523  Southern Metropolitan   
88       1900  Port Phillip   -37.8422    144.9554  Southern Metropolitan   
92       1880  Port Phillip   -37.8415    144.9585  Southern Metropolitan   
93       1970  Port Phillip   -37.8436    144.9557  Southern Metropolitan   

    Propertycount         Coordinates  
80           3280  -37.8481, 144.9499  
85           3280  -37.8447, 144.9523  
88           3280  -37.8422, 144.9554  
92           3280  -37.8415, 144.9585  
93           3280  -37.8436, 144.9557  

[5 rows x 23 columns]

In [150]:
#Примечание. В результате выполнения фильтрации возвращается новый DataFrame, полученный из исходного, 
# при этом исходная таблица melb_data остаётся без изменений.

#Также вовсе не обязательно заносить маску в отдельную переменную — можно сразу вставлять условие в операцию индексации DataFrame, например:

melb_data[melb_data['Price'] > 2000000].head()

index       Suburb               Address  Rooms Type      Price Method  \
80     80  Albert Park  112 Beaconsfield Pde      3    h  2850000.0     PI   
85     85  Albert Park     104 Richardson St      4    h  2300000.0      S   
88     88  Albert Park        29 Faussett St      2    h  2120000.0      S   
92     92  Albert Park           2 Dundas Pl      3    h  2615000.0      S   
93     93  Albert Park          23 Finlay St      5    h  2100000.0      S   

     SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
80    Buxton   4/03/2017       3.3  ...    0     211.0         198.0   
85  Marshall   7/05/2016       3.3  ...    1     153.0         180.0   
88      Greg  10/09/2016       3.3  ...    1     199.0         107.0   
92    Cayzer  10/12/2016       3.3  ...    1     177.0         181.0   
93      Greg  10/12/2016       3.3  ...    1     237.0         126.0   

    YearBuilt   CouncilArea  Lattitude  Longtitude             Regionname  \
80       1890  Port Phillip   -37.8481    144.9499  Southern Metropolitan   
85       1880  Port Phillip   -37.8447    144.9523  Southern Metropolitan   
88       1900  Port Phillip   -37.8422    144.9554  Southern Metropolitan   
92       1880  Port Phillip   -37.8415    144.9585  Southern Metropolitan   
93       1970  Port Phillip   -37.8436    144.9557  Southern Metropolitan   

    Propertycount         Coordinates  
80           3280  -37.8481, 144.9499  
85           3280  -37.8447, 144.9523  
88           3280  -37.8422, 144.9554  
92           3280  -37.8415, 144.9585  
93           3280  -37.8436, 144.9557  

[5 rows x 23 columns]

In [151]:
#Найдём количество зданий с тремя комнатами. Для этого отфильтруем таблицу по условию: 
# обратимся к результирующей таблице по столбцу Rooms и найдём число строк в ней с помощью атрибута shape:

melb_data[melb_data['Rooms'] == 3].shape[0]

5881

In [157]:
#Условия можно комбинировать, используя операторы & (логическое И) и | (логическое ИЛИ). Условия при этом заключаются в скобки.

#Усложним прошлый пример и найдём число трёхкомнатных домов с ценой менее 300 тысяч:

melb_data[(melb_data['Rooms'] == 3) & (melb_data['Price'] < 300000)].shape[0]

3

In [159]:
#Таких зданий оказалось всего три. Немного «ослабим» условие: 
# теперь нас будут интересовать дома с ценой менее 300 тысяч, у которых либо число комнат равно 3 
# либо площадь домов более 100 квадратных метров:

melb_data[((melb_data['Rooms'] == 3) | (melb_data['BuildingArea'] > 100)) & (melb_data['Price'] < 300000)].shape[0]

68

Примечание. Обратите внимание, что использование привычных операторов and и or будет неверным и приведёт к ошибке, так как они выполняют логические операции между двумя булевыми числами. В нашем случае слева и справа от оператора стоят маски (объекты Series), для которых логическую операцию надо совершить поэлементно, а операторы and и or для такого не предназначены.

In [160]:
#Фильтрацию часто сочетают со статистическими методами. 
# Давайте найдём максимальное количество комнат в таунхаусах. Так как в результате фильтрации получается DataFrame, 
# то обратимся к нему по столбцу Rooms и найдём максимальное значение:

melb_data[melb_data['Type'] == 't']['Rooms'].max()

5

In [165]:
#А теперь более сложный трюк: найдём медианную площадь здания у объектов, чья цена выше средней. 
# Для того чтобы оградить наш код от нагромождений, предварительно создадим переменную со средней ценой:

mean_price = melb_data['Price'].mean()
melb_data[melb_data['Price'] > mean_price]['BuildingArea'].median()

126.0

Фильтрация находит применение в очистке данных.

Под очисткой данных понимается удаление из данных аномальных значений (выбросов), пропусков и данных, которые не несут информацию.

Причин появления таких дефектов в данных множество: неправильная передача по сети, неверная выгрузка из базы, человеческий фактор и т. д.

→ Проблема аномальных значений состоит в том, что если мы обучаем модель на данных, содержащих выбросы, то мы имеем немалый шанс получить менее верный прогноз, чего нам, конечно же, не хотелось бы.

In [170]:
#В каком районе Мельбурна чаще всего продаются виллы и коттеджи (тип здания — h) с ценой меньше трёх миллионов?
melb_data['Regionname'][(melb_data['Price']<3000000) & (melb_data['Type']=='h')].mode()

0    Northern Metropolitan
Name: Regionname, dtype: object